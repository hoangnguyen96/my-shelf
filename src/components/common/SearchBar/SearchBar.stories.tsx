import { Meta, StoryFn } from "@storybook/react";
import { ChakraProvider } from "@chakra-ui/react";
import SearchBar from ".";
import theme from "@app/themes";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
import { mockRouter } from "@app/__mocks__/storybook";

const meta: Meta<typeof SearchBar> = {
  component: SearchBar,
  decorators: [
    (Story: StoryFn) => (
      <ChakraProvider theme={theme}>
        <RouterContext.Provider value={mockRouter}>
          <Story />
        </RouterContext.Provider>
      </ChakraProvider>
    ),
  ],
};

export default meta;

const Template: StoryFn<typeof SearchBar> = (args) => <SearchBar {...args} />;

export const Default = Template.bind({});
Default.args = {
  placeholder: "Search...",
};
