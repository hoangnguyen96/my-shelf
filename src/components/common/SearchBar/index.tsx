"use client";

import { memo, useEffect, useState } from "react";
import { usePathname, useRouter } from "next/navigation";
import {
  Input,
  InputGroup,
  IconButton,
  Select,
  Flex,
  InputProps,
  Box,
} from "@chakra-ui/react";
import { SearchIcon } from "@chakra-ui/icons";
import { getFirstPath } from "@app/utils";

const SearchBar = ({ placeholder = "Search...", ...rest }: InputProps) => {
  const [prevPath, setPrevPath] = useState("");
  const [searchTerm, setSearchTerm] = useState("");
  const [searchType, setSearchType] = useState("");
  const router = useRouter();
  const pathname = usePathname();

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleTypeChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setSearchType(event.target.value);
  };

  const handleSearch = () => {
    let part = getFirstPath(pathname);

    if (!searchTerm && !searchType) {
      return router.replace(part);
    }

    const newPath = `${part}${
      searchType && searchTerm ? `/${searchType}/${searchTerm}` : ""
    }`;

    setPrevPath(part);
    return router.replace(newPath);
  };

  useEffect(() => {
    if (!pathname?.includes(prevPath)) {
      setSearchType("");
      setSearchTerm("");
    }
  }, [pathname]);

  return (
    <Flex
      boxShadow="0 0 3px 0px #a9a9a9"
      borderRadius="40px"
      w="100%"
      maxW={540}
    >
      <Select
        aria-label="select-label"
        placeholder="Select"
        border="none"
        value={searchType}
        borderLeftRadius="40px"
        w="120px"
        h="50px"
        bgColor="backgroundTitle"
        _focusVisible={{ borderColor: "transparent" }}
        onChange={handleTypeChange}
      >
        <Box as="option" value="title">
          Title
        </Box>
        <Box as="option" value="author">
          Author
        </Box>
      </Select>
      <InputGroup size="md">
        <Input
          placeholder={placeholder}
          value={searchTerm}
          onChange={handleSearchChange}
          pr="18px"
          h="50px"
          border="none"
          borderRightRadius="40px"
          borderLeftRadius="inherit"
          _placeholder={{ color: "gray.500" }}
          {...rest}
        />
        <IconButton
          aria-label="Search"
          data-testid="submit-search"
          h="100%"
          minW="60px"
          icon={<SearchIcon w="20px" height="20px" />}
          position="absolute"
          right="0"
          top="0"
          bottom="0"
          onClick={handleSearch}
          variant="outline"
          border="none"
          outline="none"
          boxShadow="none"
          bg="transparent"
          _hover={{ bg: "transparent" }}
        />
      </InputGroup>
    </Flex>
  );
};

export default memo(SearchBar);
