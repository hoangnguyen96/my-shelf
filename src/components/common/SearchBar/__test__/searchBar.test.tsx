import "@testing-library/jest-dom";
import { fireEvent, render } from "@testing-library/react";
import SearchBar from "..";
import { useRouter } from "next/navigation";

// Mock useRouter
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

describe("SearchBar", () => {
  it("Should render correctly snapshot", () => {
    expect(render(<SearchBar placeholder="search" />)).toMatchSnapshot();
  });

  it("Should updates search term correctly", () => {
    const { getByPlaceholderText } = render(<SearchBar />);
    const inputElement = getByPlaceholderText("Search...");

    fireEvent.change(inputElement, {
      target: { value: "New search term" },
    });

    expect(inputElement).toHaveValue("New search term");
  });

  it("Should updates search type correctly", () => {
    const { getByRole } = render(<SearchBar />);

    fireEvent.change(getByRole("combobox"), {
      target: { value: "author" },
    });

    expect(getByRole("combobox")).toHaveValue("author");
  });

  it("calls router.push with correct query parameters on search", () => {
    const mockPush = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });

    const { getByPlaceholderText, getByRole } = render(<SearchBar />);
    const inputElement = getByPlaceholderText("Search...");
    const combobox = getByRole("combobox");

    fireEvent.change(inputElement, {
      target: { value: "Book" },
    });
    fireEvent.change(combobox, {
      target: { value: "title" },
    });

    fireEvent.click(getByRole("button", { name: /search/i }));

    expect(mockPush).toHaveBeenCalledWith("?type=title&query=Book");
  });
});
