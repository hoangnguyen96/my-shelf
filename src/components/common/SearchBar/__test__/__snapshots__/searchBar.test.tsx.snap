// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Should render correctly snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-19jt6ar"
      >
        <div
          class="chakra-select__wrapper css-18a3ys"
        >
          <select
            class="chakra-select css-oea4np"
          >
            <option
              value=""
            >
              Select
            </option>
            <option
              class="css-0"
              value="title"
            >
              Title
            </option>
            <option
              class="css-0"
              value="author"
            >
              Author
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-1ufd936"
            placeholder="search"
            value=""
          />
          <button
            aria-label="Search"
            class="chakra-button css-xia019"
            data-testid="submit-search"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-1k92xky"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-19jt6ar"
    >
      <div
        class="chakra-select__wrapper css-18a3ys"
      >
        <select
          class="chakra-select css-oea4np"
        >
          <option
            value=""
          >
            Select
          </option>
          <option
            class="css-0"
            value="title"
          >
            Title
          </option>
          <option
            class="css-0"
            value="author"
          >
            Author
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-162mkon"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          class="chakra-input css-1ufd936"
          placeholder="search"
          value=""
        />
        <button
          aria-label="Search"
          class="chakra-button css-xia019"
          data-testid="submit-search"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1k92xky"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
